tt_eager/tensor/tensor.cpp

```diff
@@ -116,6 +142,94 @@ void Tensor::deallocate(bool force) {
+    std::visit([this] (auto&& storage) {
+        using StorageType = std::decay_t<decltype(storage)>;
+        if constexpr(std::is_same_v<StorageType, OwnedStorage> or std::is_same_v<StorageType, DeviceStorage>) {
+            std::get<StorageType>(this->tensor_attributes->storage).buffer = storage.buffer;
+        } else if constexpr(std::is_same_v<StorageType, MultiDeviceHostStorage> or std::is_same_v<StorageType, MultiDeviceStorage>) {
+            std::get<StorageType>(this->tensor_attributes->storage).buffers = storage.buffers;
+            std::get<StorageType>(this->tensor_attributes->storage).shapes = storage.shapes;
+        }
+    }, other.get_storage()); // Non blocking storage query, since this is done for tensors that get created inside the worker thread
+    // Populate remaining MD
+    this->set_shape(other.get_shape());
+    this->set_dtype(other.get_dtype());
+    this->set_layout(other.get_layout());
+    this->tensor_attributes->metadata_populated = true;
+}
+
+std::vector<Device*> Tensor::get_workers(bool blocking) const {
+    bool wait_for_tensor_populated = (Device::get_worker_mode() == Device::WorkerQueueMode::SYNCHRONOUS) or blocking;
+    TT_FATAL(wait_for_tensor_populated or (this->workers.size()), "Worker Handles for tensor must be populated or blocking = true must be set in get_workers().");
+    if (not this->workers.size()) {
+        if (wait_for_tensor_populated) {
+            while (not this->metadata_populated());
+            std::vector<Device*> workers = {};
+            std::visit(
+            [&] (auto&& storage) {
+                using StorageType = std::decay_t<decltype(storage)>;
+                if constexpr (std::is_same_v<StorageType, DeviceStorage>) {
+                    workers = {this->device()};
+                }
+                else if constexpr (std::is_same_v<StorageType, MultiDeviceStorage>) {
+                    workers.reserve(storage.buffers.size());
+                    for (const auto& buffer : storage.buffers) {
+                        workers.push_back(buffer->device());
+                    }
+                }
+            }, this->get_storage());
+            return workers;
+        }
+    }
+    return this->workers;
+}
+
+// Getters - Spin until tensor is populated before querying tensor metadata
+const Tensor::TensorAttributes& Tensor::get_attr() const {
+    while (not this->metadata_populated());
+    return *tensor_attributes;
+}
+
+const Shape& Tensor::get_legacy_shape() const {
+    return this->get_attr().shape.value();
+}
+
+const ttnn::Shape& Tensor::get_shape() const {
+    return this->get_attr().shape;
+}
+const DataType& Tensor::get_dtype() const {
+    return this->get_attr().dtype;
+}
+const Layout& Tensor::get_layout() const {
+    return this->get_attr().layout;
+}
+
+const Storage& Tensor::get_storage() const {
+    // Per device bool for storage population for multidevice
+    return this->get_attr().storage;
+}
+
```

Comment:
Looks clean while handling the sync necessary to retrieve asynchronously populated attributes!